src/sss_client/nss_mc_common.c

```diff
@@ -130,25 +130,25 @@ errno_t sss_nss_check_header(struct sss_cli_mc_ctx *ctx)
 
 static void sss_nss_mc_destroy_ctx(struct sss_cli_mc_ctx *ctx)
 {
-    uint32_t active_threads = ctx->active_threads;
-#if HAVE_PTHREAD
-    pthread_mutex_t *mutex = ctx->mutex;
-#endif
 
     if ((ctx->mmap_base != NULL) && (ctx->mmap_size != 0)) {
         munmap(ctx->mmap_base, ctx->mmap_size);
     }
+    ctx->mmap_base = NULL;
+    ctx->mmap_size = 0;
+
     if (ctx->fd != -1) {
         close(ctx->fd);
     }
-    memset(ctx, 0, sizeof(struct sss_cli_mc_ctx));
     ctx->fd = -1;
 
-    /* restore count of active threads */
-    ctx->active_threads = active_threads;
-#if HAVE_PTHREAD
-    ctx->mutex = mutex;
-#endif
+    ctx->seed = 0;
+    ctx->data_table = NULL;
+    ctx->dt_size = 0;
+    ctx->hash_table = NULL;
+    ctx->ht_size = 0;
+    ctx->initialized = UNINITIALIZED;
```

Comment:
Hi,

I think this change fixes the race condition. My concern is only that if `struct sss_cli_mc_ctx` gets new members in future there is a fair chance the the corresponding initialization might be forgotten here. Maybe the initialization can be done with a macro which is defined in `nss_mc.h` near the structure definition and `SSS_CLI_MC_CTX_INITIALIZER` to lower this risk?

bye,
Sumit

Reply:
I was thinking about it, but:
 - how to use initializer list in assignment? `memcpy` of a local stub variable?
 - this will have the same issue, as original code - `mutex` and `active_threads` will be touched

I also guess that when `ctx->initialized == UNINITIALIZED` other fiels (besides two mentioned) don't actually matter much, but I didn't check (and can't predict about potential future fields, of course).

Reply:
Looks good to me.

Can we address @sumit-bose's concerns with appropriate comment/warning next to the structure declaration?

Something like
/* if you extend this structure, do not forget update the initialization code (simple memset caused race condition in mutex handling)... */
?

Reply:
Hi,

I was thinking of a macro like e.g.

```
#define MC_RESET(ctx) do { \
    (ctx)->initialized = UNINITIALIZED; \
    (ctx)->fd = -1; \
    .... \
} while (0);
```

But I'm fine with a comment as Tomas suggested as well.

bye,
Sumit

Reply:
Added a comment.
@sumit-bose, @thalman, please take a look.

Reply:
Sorry for nitpicking, I'm not native speaker but I think it should be `structure _is_ extended` or `structure extension`?

Reply:
Updated.